README
This is project libodb, created on Wed Mar 22 15:43:28 CET 2006.


ChangeLog
2007-10-25  Morten Kjeldgaard  <mok@bioxray.dk>

	*  Bumped version to 1.0.2
	*  Added README file in config/. 
	*  Now uses $top_builddir variable in testing/Makefile.am
	*  Add fink .info file
	*  Fixed compilation problems in Ubuntu 7.10, need to include <cassert>.
	*  .spec file updated, package now split in devel files and shared
	*  library.
	*  Add GPL3 clause to all source files.
	
2006-09-07  Morten Kjeldgaard  <mok@ghost.bioxray.dk>

	* libodb.spec.in (BuildRoot): Fixed a couple of bugs, and added 
	  two jiffies, odbcat and odbdir. 

2006-03-22  Morten Kjeldgaard  <mok@bioxray.dk>

	* First version of rpm package.
	
Local variables:
mode: change-log
mode: font-lock
End:



testing/aaa.cpp
// field width
#include <iostream>
using namespace std;

int main () {
  cout << 100 << endl;
  cout.width(10);
  cout << 100 << endl;
  cout.fill('x');
  cout.width(15);
  cout.setf(ios::left);
  cout << 100 << endl;
  return 0;
}



testing/bbb.cpp
// example on ostream::operator<<
#include <iostream>
using namespace std;

int main () {

  char str[] = "Test sentence";
  int val = 65;
  char ch = 'A';

  cout << str << endl;     // Insert string

  cout << ch << endl;      // Insert single character

  cout.width (10);
  cout << right;           // Insert manipulator

  cout << val << endl;     // Insert integer

  return 0;
}


testing/sss.cpp
#include <iostream>
#include <string>

using std::cerr;
using std::endl;

main () 
{
  int i;
  int j;
  std::string x;

  i = 45;
  j = 123;
  x = "Morten";

  cerr.width(25);
  cerr << x ;
  cerr << "#" ;
  cerr.width(10);
  cerr << i << "#";
  cerr.width(5);
  cerr << j << "#";
  cerr << endl;

}


src/odb.h
/*
    C++ class to deal with an O database (collection of datablocks)
    Copyright (C) 2003-2007 Morten Kjeldgaard

    This program is free software: you can redistribute it and/or
    modify it under the terms of the GNU Lesser General Public License
    as published by the Free Software Foundation, either version 3 of
    the License, or (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include <map>
#include <string>

#include "datablock.h"

#ifndef _odb_h
#define _odb_h

typedef std::map<std::string, class ODatablock *, std::less<std::string> > maptype;


class ODatabase {
private:
  std::string _path;
  std::string _name;
  maptype _db;

public:
  // Constructors
  ODatabase(void);
  ODatabase(std::string);

  // Destructor
  ~ODatabase(void);

  // Utils...
  void open(std::string);
  void get (void);		 // suck in a file, binary or formatted ...
  void write(std::string path);		 // we only write out formatted...
  // NOTE: Overload << and >> !!!!!!!!!

  void dir (void);                    // directory...

  ODatablock *fetch(std::string);

};

#endif

// Local variables:
// mode: c++
// mode: font-lock
// fill-column: 75
// comment-column: 50
// End:


src/odb.cpp
/*
    C++ class to deal with an O database (collection of datablocks).
    Copyright (C) 2003-2007 Morten Kjeldgaard

    This program is free software: you can redistribute it and/or
    modify it under the terms of the GNU Lesser General Public License
    as published by the Free Software Foundation, either version 3 of
    the License, or (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/
//   -*- mode: c++; mode: font-lock; fill-column: 75; comment-column: 50; -*-
//
// $Id: 
//

#include "odb.h"
#include "datablock.h"
#include "ofile.h"
#include <cstdlib>
#include <iostream>
#include <iterator>

using std::cout;
using std::cerr;
using std::endl;

// Constructor 1
ODatabase::ODatabase(void) {}

// Constructor 2
ODatabase::ODatabase(std::string path) 
{
  open(path);
}

// Destructor
ODatabase::~ODatabase(void)
{
  // Here we need to zap all datablocks!
  // Create an iterator for the _db map and delete all
  // the datablocks (they have been created with 'new')
}

// List contents of the database
void ODatabase::dir(void)
{
  maptype::iterator it;

  it = _db.begin();
  while (it != _db.end()) {

    cout.flush();
    cout << std::left;
    cout.width(27);
    cout << it->first;
    cout.width(3);
    cout << it->second->get_type();
    cout << std::right;
    cout.width(10);
    cout << it->second->size();
    cout.width(10);
    cout << it->second->get_byte_size();
    cout << endl;

    it++;
  }
}

// Write database to (formatted) file...
void ODatabase::write(std::string path)
{
  // bla bla...
}


void ODatabase::open(std::string path)
{
  int i = _path.find_last_of('/');     // last occurrence of '/'
  _path = path;
  if (i != std::string::npos)
    _name = _path.substr(i); 	      // substring from i to eos
  else
    _name = path;
}


// Populate database from O format file
void ODatabase::get() 
{
  char nam[ODB_DBNAME_SIZE], fmt[ODB_FMTSIZ], typ, *s;
  int siz;
  void *vect, *data;
  OFile ofile;

  if (!ofile.open(_path))
    return;

  while (ofile.get_header(nam, typ, siz, fmt)) {

    // strip spaces off end of datablock name 
    s = &nam[ODB_DBNAME_SIZE - 2];
    while (*s <= 32 && s > nam)
      *s-- = '\0';
    nam[ODB_DBNAME_SIZE-1] = '\0';

    switch(typ) {

    case 'I':
      {
	IntBlock *ib;
	ib = new IntBlock(siz);

	ib->_data.resize(siz);
	ofile.get_intblock (&ib->_data[0], siz);
	ib->set_name(nam);
	_db[std::string(nam)] = ib;
	//cerr << ib->_data[0] << endl;
      }
      break;

    case 'R':
      {
	RealBlock *rb;
	rb = new RealBlock(siz);

	rb->_data.resize(siz);
	ofile.get_realblock (&rb->_data[0], siz);
	rb->set_name(nam);
	_db[std::string(nam)] = rb;
      }
      break;
      
    case 'C':
      {
	CharBlock *cb;
	char buf[8], *ch;
	cb = new CharBlock(siz);
	buf[6] = 0;
	s = new char[siz*6*sizeof(char)];  
	ofile.get_charblock (s, siz, fmt);
	cb->set_name(nam);
	for (int i=0; i < siz; i++) {
	  memcpy(buf, &s[6*i], 6);
	  // strip spaces off end 
	  ch = &buf[6];
	  while (*ch <= 32 && ch > buf)
	    *ch-- = '\0';
	  cb->_data.push_back(std::string(buf));
	}
	delete [] s;
	_db[std::string(nam)] = cb;
      }			  // add to dictionary
      break;

    case 'T':
      {
	TextBlock *tb;
	char *ch, *t;
	int bufsiz;
	int nrec = 0;

	tb = new TextBlock(siz);
	bufsiz = siz*73*sizeof(char); // O text blocks have max 73 chars per line
	s = new char [bufsiz];
	t = new char [bufsiz];
	ofile.get_textblock (s, siz);
	tb->set_name(nam);

	// extract the individual strings into 't'
	for (int i=0, j=0; i<bufsiz; i++) { 
	  t[j] = s[i];
	  if (t[j] == '\000')
	    break;
	  if (t[j] == '\r') {
	    ch = &t[j];
	    while (*ch <= 32 && ch > t) // strip spaces off end
	      *ch-- = '\0';

	    // create string and store
	    tb->_data.push_back(std::string(t));
	    nrec++;
	    j=0;
	  } else {
	    j++;
	  }
	}
	tb->set_size(nrec);
	delete [] t;
	delete [] s;
	_db[std::string(nam)] = tb;		  // add to dictionary
      }
      break;

    } // end switch (typ)
  }
  ofile.close();
  return;
}

ODatablock *ODatabase::fetch(std::string nam) {
  
  return _db[nam];
}

// Local variables:
// mode: c++
// mode: font-lock
// fill-column: 75
// comment-column: 50
// End:


src/ofile.h
/*
    Declare member functions for the Ofile class that deals with files
    in O format, either formatted or binary. 
    Copyright (C) 2003-2007 Morten Kjeldgaard

    This program is free software: you can redistribute it and/or
    modify it under the terms of the GNU Lesser General Public License
    as published by the Free Software Foundation, either version 3 of
    the License, or (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#ifndef _ofile_h
#define _ofile_h
#include <string>
#include <fstream>

#define ODB_FMTSIZ 64

class OFile {

 private:
  std::string _path;
  bool _binary;					  // true if file is binary
  bool _swap;				          // true if bytes must be swapped
  std::ifstream _file;				  // file descriptor 
  int binfil(void);

 public:
  OFile( void );				  // constructor
  OFile( std::string );				  // constructor
  ~OFile();					  // destructor

  // Misc member functions...
  bool open(std::string);
  void close(void);
  bool get_header ( char*, char &, int &, char * );
  bool get_intblock ( int *, int );
  bool get_realblock ( float *, int );
  bool get_charblock ( char *, int, char * );
  bool get_textblock( char *, int &);
};
#endif

// Local variables:
// mode: c++
// mode: font-lock
// fill-column: 75
// comment-column: 50
// End:


src/datablock.h
/*
    C++ class to deal with O datablocks.
    Copyright (C) 2003-2007 Morten Kjeldgaard

    This program is free software: you can redistribute it and/or
    modify it under the terms of the GNU Lesser General Public License
    as published by the Free Software Foundation, either version 3 of
    the License, or (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#ifndef _datablock_h
#define _datablock_h

#include <iostream>
#include <vector>
#include <string>
#include <cassert>

// Length of parameter block names
#define ODB_DBNAME_SIZE 26


// An O "character" variable is 5 bytes.
// TODO! Overload the [] operator!

typedef struct ochar {
  char c[6];
} ochar;


// Datablock base class 
class ODatablock {
protected:
  char _name[ODB_DBNAME_SIZE];
  char _type;
  int _size;
  int _bytes;

public:

  // Constructor
  ODatablock(void) { _size = -1; _type = '*'; _bytes = 0; };

  // Destructor
  virtual ~ODatablock(void) {};

  virtual void set_name(char * nam) {
    memcpy (_name, nam, ODB_DBNAME_SIZE);
    return;
  };

  virtual char *get_name() { return _name; };
  virtual char get_type() { return _type; };
  virtual int size () = 0; 
  virtual int capacity () = 0; 
  virtual int get_byte_size () = 0; 
  virtual void write () = 0;

};

// Class describing an integer datablock

class IntBlock : public ODatablock {

public:
  std::vector<int> _data; 
  IntBlock(void) {_type = 'I';};
  IntBlock(int size) { _type = 'I'; _data.reserve(size); _size = size; };
  ~IntBlock(void) { _data.clear();};
  
  // set size of the block
  void set_size(int size)
  {
    if (size > 0) {
      _size = size;
      _data.reserve(size);
    }
  };

  // return the size
  int size()
  {
    int s = _data.size();
    assert (s == _size);
    return _size;
  }

  // return allocated size (number of elements)
  int capacity() { return _data.capacity();}

  // return the size in bytes of this block
  int get_byte_size () { 
    if (_bytes == 0) {
      _bytes = _data.size()*sizeof(int); 
    }
    return _bytes;
  }

  // return the data
  std::vector<int> &fetch_data() { return _data; }


  // write the datablock to stdout
  void write()
    {
      std::cout << _name << " " << _type << " " << _size << "(i10)" << std::endl;
      for (int i=0; i < _size; i++) {
	std::cout << _data[i] << std::endl;
      }
    }
};


// Class describing a real datablock

class RealBlock : public ODatablock {

public:
  std::vector<float> _data; 
  RealBlock(void){ _type = 'R'; };
  RealBlock(int size){ _type = 'R'; _data.reserve(size); _size = size; };
  ~RealBlock(void){ _data.clear();};

  // set size of the block
  void set_size(int size)
  {
    if (size > 0) {
      _size = size;
      _data.reserve(size);
    }
  };

  // return the allocated size
  int size()
  {
    int s = _data.size();
    assert (s == _size);
    return _size;
  }

  // return allocated size (number of elements)
  int capacity()   {return _data.capacity();}

  // return the size in bytes of this block
  int get_byte_size () { 
    if (_bytes == 0) {
      _bytes = _data.size()*sizeof(float);
    }
    return _bytes;
  }

  // return the data
  std::vector<float> &fetch_data() { return _data; }

  // write the datablock to stdout
  void write()
    {
      std::cout << _name << " " << _type << " " << _size << " (f10.0)" << std::endl;
      for (int i=0; i < _size; i++) {
	std::cout << _data[i] << std::endl;
      }
    }

};


// Class describing a char datablock. In O,
// these are arrays of character*6.

class CharBlock : public ODatablock {

public:
  std::vector<std::string> _data; 
  CharBlock(void){ _type = 'C'; };
  CharBlock(int size){ _type = 'C'; _data.reserve(size); _size = size; };
  ~CharBlock(void){ _data.clear();};

  // set size of the block
  void set_size(int size)
  {
    if (size > 0) {
      _size = size;
      _data.reserve(size);
    }
  };

  // return the allocated size
  int size()
  {
    int s = _data.size();
    assert (s == _size);
    return _size;
  }

  // return allocated size (number of elements)
  int capacity()   {return _data.capacity();}

  // return the size in bytes of this block
  int get_byte_size () { 
    if (_bytes == 0) {
      for (int i=0; i < _data.size(); i++)
	_bytes += _data[i].size();
    }
    return _bytes;
  }

  // return the data
  std::vector<std::string> &fetch_data() { return _data; }

  // write the datablock to stdout
  void write()
    {
      std::cout << _name << " " << _type << " " << _size << "(a6)" << std::endl;
      for (int i=0; i < _size; i++) {
	std::cout << _data[i] << std::endl;
      }
    }

};


// Class describing a text datablock.

class TextBlock : public ODatablock {

public:
  std::vector<std::string> _data; 
  TextBlock(void){ _type = 'T'; };
  TextBlock(int size){ _type = 'T'; _data.reserve(size); _size = size; };
  ~TextBlock(void){ _data.clear();};

  // set size of the block
  void set_size(int size)
  {
    if (size > -1) {
      _size = size;
      _data.reserve(size);
    }
  };


  // return the true size == the actual number of values stored..
  int size()
  {
    int s = _data.size();
    assert (s == _size);
    return s;
  }


  // return the max length of any element
  int maxlength() {
    int curmax = 0;
    for (int i=0; i < _size; i++) {
      curmax = ( curmax > _data[i].length() )? curmax : _data[i].length();
    }
    return curmax;
  }


  // return allocated size (number of elements)
  int capacity()   {return _data.capacity();}


  // return the size in bytes of this block
  int get_byte_size () { 
    if (_bytes == 0) {
      for (int i=0; i < _data.size(); i++)
	_bytes += _data[i].size();
    }
    return _bytes;
  }

  // return the data
  std::vector<std::string> &fetch_data() { return _data; }

  // write the datablock to stdout
  void write()
    {
      std::cout << _name << " " << _type << " " << _size << " " 
		<< this->maxlength() << std::endl;
      for (int i=0; i < _size; i++) {
	std::cout << _data[i] << std::endl;
      }
    }
};

#endif

// Local Variables: 
// mode: c++
// mode: font-lock
// fill-column: 75
// comment-column: 50
// End:


src/ofile.cpp
/*
    Define member functions for the Ofile class that deals with files
    in O format, either formatted or binary. 
    Copyright (C) 2003-2007 Morten Kjeldgaard

    This program is free software: you can redistribute it and/or
    modify it under the terms of the GNU Lesser General Public License
    as published by the Free Software Foundation, either version 3 of
    the License, or (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "ofile.h"
#include <iostream>
#include <cassert>
#include <cstring> // to support various string functions
#include <cstdlib> // to support strtol etc.

using std::cerr;

/*
  A simple parser for the fortran formats that the type C datablocks
  are stored in. Generate a string describing the format, so the
  program knows what to expect in the file. The function knows that 
  type C variables are 6 bytes always. Examples:

  (2a)         -> 123456123456
  (1x,5a)      -> _123456123456123456123456123456
  (5(1x,a6))   -> _123456_123456_123456_123456_123456
  (1x,2(2x,a)) -> ___123456__123456
  (aaaaa)      -> 123456123456123456123456123456

*/
#define RSIZ 256

static char *parse_format(char *fmt)
{
  register int i;
  char *result, *f, *r, *t;
  int mult = 0;

  result = new char[RSIZ];
  memset (result, 0, RSIZ);
  
  f = fmt;
  r = result;
  
  while (*f) {
    switch (*f) {
    case '(':
      t = parse_format(++f);
      if (mult == 0) mult = 1;
      for (i=0; i < mult; i++)
	strncat(result, t, RSIZ);
      delete [] t;
      return result;
      break;
    case 'x':
    case 'X':
      if (mult == 0) mult = 1;
      for (i=0; i < mult; i++)
	*r++ = '_';
      mult = 0;
      break;
    case 'a':
    case 'A':
      if (mult == 0) mult = 1;
      for (i=0; i < mult; i++) {
	*r++ = '0'+1;
	*r++ = '0'+2;
	*r++ = '0'+3;
	*r++ = '0'+4;
	*r++ = '0'+5;
	*r++ = '0'+6;
      }
      mult = 0;
      break;
    case '0':
    case '1':
    case '2':
    case '3':
    case '4':
    case '5':
    case '6':
    case '7':
    case '8':
    case '9':
      mult = mult*10+(*f)-'0';
      break;
    case ')':
      return result;
    case ',':
      mult = 0;
      break;
    }
    f++;
  }
  cerr << "parse_format: shouldn't happen\n";
  return NULL;
}


// Class constructors

OFile::OFile(void) {
}

OFile::OFile(std::string path) {
  open(path);
}

// Open the file. Determine whether file is binary or formatted. Determine
// whether or not to swap bytes. Open file and save descriptor.

bool OFile::open(std::string path) {

  _path = path;
  
  switch (binfil()) {

  case 1: 
    _binary = true;
    _file.open(_path.c_str(), std::ios::in | std::ios::binary);
    break;

  case 0:
    _binary = false;
    _file.open(_path.c_str(), std::ios::in);
    break;

  default:
    return false;
  }

  if (_file.fail()) return false;

  // determine endianism...
  union {
    unsigned int ui;
    char byte[4];
  } i;

  i.ui = 0;
  i.byte[3]= 1;

  if (i.ui == 1) {
    _swap = false; 
  }
  else {
    _swap = true;
  }
  return true;
}

// Class destructor, closes file.

OFile::~OFile (void)
{
  _file.close();
}

///  binfil -- a private class member. Attempt to open file and read in
///  the first eight bytes. Return 'true' if fnam is a binary O file, else
///  'false'. An O binary file normally has the byte pattern [0 0 0 036 .] in
///  the first 5 bytes of the file.

int OFile::binfil(void)
{
  std::ifstream File;
  char buf[8];

  File.open(_path.c_str(), std::ios::in | std::ios::binary);

  if (File.fail()) {
    cerr << "binfil: could not open file\n";
    return -1;
  }

  File.read((char *)buf, 8);
  File.close();

  if ((buf[0]&buf[1]&buf[2]) == 0 && buf[3] == 30 && buf[4] == '.') {
    // file has binary signature
    return 1;
  }
  if ((buf[0]&buf[1]&buf[2]) == 0 && buf[3] == 30) {
    // The only binary O files that do _not_ have a '.' in the fifth byte
    // are the dgnl data files
    return 1;
  }

  // file is formatted
  return 0;
}


void OFile::close(void)
{
  _file.close();
}

//  Swap bytes in n 4-byte words
static void swap4 (char *buffer, int n)
{
  char j;

  for (int i=0; i < n*4; i+=4) {
    j = buffer[i];
    buffer[i] = buffer[i+3];
    buffer[i+3] = j;
    j = buffer[i+1];
    buffer[i+1] = buffer[i+2];
    buffer[i+2] =j; 
  }
}

// In the code below we use a bunch of C string functions. Declare them here...

#include <cstring>

/// Read the datablock header.

bool OFile::get_header(char *nam, char &typ, int &siz, char *fmt) 
{

  if (_file.fail()) {
    cerr << "failed\n";
    return false;
  }

  if (_file.eof()) {
    cerr << "end of file reached\n";
    return false;
  }

  if (_binary) {
    int n;
    int rl1, rl2;

    _file.read ((char *)&rl1, sizeof(int));
    if (_file.fail()) return false;
    if (_swap) swap4 ((char *)&rl1, 1);
    _file.read(nam, 25);

    // convert datablock name to lower case
    for (n=0; n<25; n++)
      nam[n] = tolower(nam[n]);
    _file.read ((char *)&typ, sizeof (char));
    _file.read ((char *)&siz, sizeof (int));
    if (_swap) swap4 ((char *)&siz, 1);
    _file.read((char *)&rl2, sizeof (int));
    if (_swap) swap4 ((char *)&rl2, 1);
    assert (rl1 == rl2);
    return true;

  } else {

    // file is formatted.
    const int sz = 256;
    char *stat;
    char buf[sz], *cp;
    int i;
    char num[64];

    memset (buf, 0, sz);
    memset (num, 0, 64);
    memset (fmt, 0, ODB_FMTSIZ);
    memset (nam, 0, 26);

    // Get the first non-comment line.

    buf[0] = '!';
    while (buf[0] == '!' || buf[0] == '\n' || buf[0] == 0) {
      _file.getline(buf,sz);
      if (_file.eof())
	return false;
    }

    i=0;
    cp = nam;
    while (!isspace(buf[i]))			  // copy name
      *cp++ = tolower(buf[i++]);

    while (isspace(buf[i]))			  // step forward to type
      i++;
    typ = toupper(buf[i++]);			  // copy type


    while (isspace(buf[i]))			  // step forward to size
      i++;
    cp = num;
    while (!isspace(buf[i]))			  // copy size 
      *cp++ = buf[i++];
    siz = (int)strtol(num, &stat, 10);		  // decode size
    if (*stat) {
      cerr << "error decoding size\n";
      return false;
    }

    while (isspace(buf[i]))			  // step forward until fmt string
      i++;
    cp = fmt;
    while (!isspace(buf[i]) && buf[i])		  // copy fmt string
      *cp++ = buf[i++];

    return true;
  }
}


///  Read 'size' integers from the binary fortran file. Swap bytes if
///  necessary, file is always in big-endian order.

bool OFile::get_intblock(int *istore, int size) 
{
  if (_binary) {
    int rl1, rl2;

    _file.read ((char *)&rl1, sizeof(int));  
    if (_file.fail()) {
      return -1;
    }
    if (_swap) swap4 ((char *)&rl1, 1);
    _file.read ((char *)istore, rl1);
    if (_swap) swap4 ((char *)istore, size);
    _file.read ((char *)&rl2, sizeof(int));  
    if (_swap) swap4 ((char *)&rl2, 1);

    assert (rl1 == rl2);

    if (4*size != rl2) {
      cerr <<  "get_intblock: expected " << 4*size << " got " << rl2 << " elements\n";
      return false;
    }
  } else {

    // formatted file...

    char buf[64], *stat;

    for (int i=0; i<size; i++) {
      _file >> buf;
      istore[i] = (int)strtol(buf, &stat, 10); 
      if (*stat) {
	cerr <<  "non-digits in datablock\n";
	return false; 
      }
    }
  }
  return true;
}

///   Read 'size' floats from the binary fortran file. Swap bytes if
///   necessary, file is always in big-endian order.

bool OFile::get_realblock(float *rstore, int size) 
{
  if (_binary) {
    int rl1, rl2;

    _file.read ((char *)&rl1, sizeof(int));  
    if (_file.fail()) {
      return -1;
    }
    if (_swap) swap4 ((char *)&rl1, 1);
    _file.read ((char *)rstore, rl1);
    if (_swap) swap4 ((char *)rstore, size);
    _file.read ((char *)&rl2, sizeof(int));  
    if (_swap) swap4 ((char *)&rl2, 1);

    assert (rl1 == rl2);

    if (4*size != rl2) {
      cerr <<  "get_realblock: expected " << 4*size << " got " << rl2 << " elements\n";
      return false;
    }
  } else {

    // formatted file...
    char buf[64], *stat;
    
    for (int i=0; i<size; i++) {
      _file >> buf;
      rstore[i] = (float)strtod(buf, &stat); 
      if (*stat) {
	cerr <<  "non-digits in datablock\n";
	return false; 
      }
    }
  }
  return true;
}

/// Read 'size' C6 variables

bool OFile::get_charblock(char *cstore, int size, char *fmt)
{
  if (_binary) {
    int rl1, rl2;

    _file.read ((char *)&rl1, sizeof(int));  
    if (_file.fail()){
      return false;
    }
    if (_swap) swap4 ((char *)&rl1, 1);
    _file.read (cstore, rl1);
    _file.read ((char *)&rl2, sizeof(int));  
    if (_swap) swap4 ((char *)&rl2, 1);

    assert (rl1 == rl2);

    if (6*size != rl2) {
      cerr << "get_charblock: expected " << 6*size << ", got " << rl2 << " elements\n";
      return false;
    }
    return true;

  } else {

    // Formatted input
    char c=0, *a, *t, *s;
    int i, eol, inword;

    t = parse_format(fmt);
    s = t;
    a = cstore;
  
    i = 0;
    eol = 0;

    while (i < size) {
      if (!*s) { // if end of format string, rewind it.
	s = t;   // and reset everything
	eol = 0;
	while (_file.get() != '\n')
	  ;
      }	      

      inword = 1;
      while (inword) {
	if (!eol)
	  c = _file.get();
	if (c == '\n')
	  eol = 1;
	switch (*s++) {
	case '_':
	  break;
	case '1':
	case '2':
	case '3':
	case '4':
	case '5':
	  if (eol)
	    //*a++ = ' ';
	    *a++ = '#';
	  else
	    *a++ = c;
	  break;
	case '6':
	  if (eol)
	    //*a++ = ' ';
	    *a++ = '6';
	  else
	    *a++ = c;
	  i++;
	  inword = 0;
	  break;
	}
      }  // end get one word
    }
    delete [] t;
    return true;

  }
}


///  Read a text datablock. The definition of the size of a text block
/// is different in binary format, where 'size' is the number of bytes,
/// or formatted, where 'size' is the number of lines...

bool OFile::get_textblock(char *text, int &size) 
{
  if (_binary) {
    int rl1, rl2;

    _file.read ((char *)&rl1, sizeof(int));  
    if (_file.fail()) {
      return false;
    }
    if (_swap) swap4 ((char *)&rl1, 1);
    _file.read (text, rl1);
    _file.read ((char *)&rl2, sizeof(int));  
    if (_swap) swap4 ((char *)&rl2, 1);

    assert (rl1 == rl2);

    if (size != rl2) {
      cerr << "get_textblock: expected " << size <<", got" << rl2 << " bytes\n";
    }

    // count no of lines and return in size
    size = 0;
    for (int i=0; i<rl2; i++)
      if (text[i] == '\r')
	size++;

  } else {

    // formatted text block...

    char buf[256];
    int j, n;

    j = 0;
    for (int i=0; i<size; i++) {
      _file.getline (buf, 256);
      n = strlen(buf);
      buf[n++] = '\r';				  // like in binary file
      strncpy (text+j, buf, n);
      j += n;
    }
    text[j] = '\000';
  }
  return true;
}

// Local variables:
// mode: c++
// mode: font-lock
// fill-column: 75
// comment-column: 50
// End:
